### PR TITLE
Feature: Structure block translation

### PR DESCRIPTION
**Pull request by onebeastchris**
Builds on https://github.com/GeyserMC/Geyser/pull/2044 - huge thanks to Camotoy for that.

Current issues/TO-DO's:
- [x] fix offset application when we're changing rotations
- [x] fix offsets when we're mirroring *and* rotating (when looking at structure block loading selections)
- [x] ensure our added offsets are properly removed before sending them over to the Java server
- [x] yeet debug

Opening as a draft in case someone wants to test this :p

### CODE REVIEW

**Comment by Tydium** [comment]
I love you

**Comment by onebeastchris** [comment]
I'm off to vacation so i can't continue this PR for a week. Currently, rotations combined with mirroring does work, but modifying the rotation/mirror or opening a structure block with a mirror/rotation applied on Bedrock causes havoc with the offsets - in other words, offset removing isn't quite working. Getting close though :p

File: core/src/main/java/org/geysermc/geyser/session/cache/StructureBlockCache.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ */
+
+package org.geysermc.geyser.session.cache;
+
+import lombok.Getter;
+import lombok.Setter;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.cloudburstmc.math.vector.Vector3i;
+
+@Setter @Getter
```
Comment (rtm516) [inline core/src/main/java/org/geysermc/geyser/session/cache/StructureBlockCache.java:33]: Do we not normally keep these on separate lines?
```suggestion
@Setter
@Getter
```

File: core/src/main/java/org/geysermc/geyser/translator/level/block/entity/StructureBlockBlockEntityTranslator.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            UpdateBlockPacket spawnerBlockPacket = new UpdateBlockPacket();
+            spawnerBlockPacket.setDataLayer(0);
+            spawnerBlockPacket.setBlockPosition(position);
+            spawnerBlockPacket.setDefinition(session.getBlockMappings().getStructureBlockFromMode(mode));
+            session.sendUpstreamPacket(spawnerBlockPacket);
+        }
+
+        return super.getBlockEntityTag(session, type, x, y, z, tag, blockState);
+    }
+    @Override
```
Comment (rtm516) [inline core/src/main/java/org/geysermc/geyser/translator/level/block/entity/StructureBlockBlockEntityTranslator.java:71]:
```suggestion

    @Override
```

File: core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureBlockUpdateTranslator.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            default -> UpdateStructureBlockMode.SAVE;
+        };
+
+        StructureMirror mirror = switch (data.getSettings().getMirror()) {
+            case X -> StructureMirror.FRONT_BACK;
+            case Z -> StructureMirror.LEFT_RIGHT;
+            default -> StructureMirror.NONE;
+        };
+
+        com.github.steveice10.mc.protocol.data.game.level.block.StructureRotation rotation = switch (settings.getRotation()) {
```
Comment (rtm516) [inline core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureBlockUpdateTranslator.java:72]:
```suggestion
        StructureRotation rotation = switch (settings.getRotation()) {
```

File: core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureTemplateDataRequestTranslator.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    StructureBlockUtils.sendStructureData(session, size, packet.getName());
+                    return;
+                }
+            }
+
+            // Request a "structure load" from Java server, so it sends us the structure's size
+            // See the block entity translator for more info
+            session.getStructureBlockCache().setCurrentStructureBlock(packet.getPosition());
+
+            com.github.steveice10.mc.protocol.data.game.level.block.StructureRotation rotation = switch (settings.getRotation()) {
```
Comment (rtm516) [inline core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureTemplateDataRequestTranslator.java:82]:
```suggestion
            StructureRotation rotation = switch (settings.getRotation()) {
```

File: core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureTemplateDataRequestTranslator.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import org.geysermc.geyser.util.StructureBlockUtils;
+
+/**
+ * Packet used in Bedrock to load structure size into the structure block GUI. It is sent every time the GUI is opened.
+ * Or, if the player updates the structure name. Which we can use to request the structure size from the Java server!
+ * <p>
+ * Java does not have this preview, instead, Java clients are forced out of the GUI to look at the area.
+ */
+@Translator(packet = StructureTemplateDataRequestPacket.class)
+public class BedrockStructureTemplateDataRequestTranslator extends PacketTranslator<StructureTemplateDataRequestPacket> {
```
Comment (rtm516) [inline core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockStructureTemplateDataRequestTranslator.java:50]: This feels very similar to BedrockStructureBlockUpdateTranslator, could some of this be moved into a shared class to save code duplication?

**Review by rtm516 — CHANGES_REQUESTED** [review]

**Review by Kas-tle — APPROVED** [review]

**Review by rtm516 — APPROVED** [review]

**Pull request merged**